File: app/src/main/kotlin/tech/mmarca/openvitals/features/manualentry/activity/ActivityEntryTypes.kt
package tech.mmarca.openvitals.features.manualentry.activity

import tech.mmarca.openvitals.features.manualentry.*
import tech.mmarca.openvitals.features.manualentry.activity.*
import tech.mmarca.openvitals.features.manualentry.activity.recording.*
import tech.mmarca.openvitals.features.manualentry.activity.routeimport.*
import tech.mmarca.openvitals.features.manualentry.body.*
import tech.mmarca.openvitals.features.manualentry.hydration.*
import tech.mmarca.openvitals.features.manualentry.mindfulness.*
import tech.mmarca.openvitals.features.manualentry.vitals.*



import androidx.annotation.StringRes
import androidx.health.connect.client.records.ExerciseSegment
import androidx.health.connect.client.records.ExerciseSessionRecord
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.domain.model.PlannedExerciseCompletion
import tech.mmarca.openvitals.domain.model.PlannedExerciseData
import tech.mmarca.openvitals.domain.model.activeSteps

enum class ActivityRecordingSensor {
    NONE,
    GPS,
    BLE,
    PROXIMITY,
    ACCELEROMETER,
    STEP_DETECTOR,
}

enum class ActivityRepetitionUnit {
    REPETITIONS,
    STEPS,
}

data class ActivityEntryType(
    val exerciseType: Int,
    val id: String = exerciseType.toString(),
    @param:StringRes val labelRes: Int,
    val supportsGpsRoute: Boolean = true,
    val supportsDistance: Boolean = true,
    val supportsElevation: Boolean = false,
    val recordingSensor: ActivityRecordingSensor = if (supportsGpsRoute) {
        ActivityRecordingSensor.GPS
    } else {
        ActivityRecordingSensor.NONE
    },
    val segmentType: Int? = null,
    val defaultTitle: String? = null,
    val repetitionUnit: ActivityRepetitionUnit? = null,
)

val ActivityEntryType.supportsLiveRecording: Boolean
    get() = recordingSensor != ActivityRecordingSensor.NONE

val ActivityEntryType.supportsSetRepetitions: Boolean
    get() = repetitionUnit == ActivityRepetitionUnit.REPETITIONS

val ActivityEntryType.isRepetitionLike: Boolean
    get() = repetitionUnit != null

val ActivityEntryType.supportsStepCounting: Boolean
    get() = repetitionUnit == ActivityRepetitionUnit.STEPS

/** Generic set types (calisthenics, strength sets): every step picks its own exercise. */
val ActivityEntryType.supportsMixedExercises: Boolean
    get() = supportsSetRepetitions && recordingSensor == ActivityRecordingSensor.NONE

fun activityEntryTypeById(id: String?): ActivityEntryType? =
    id?.let { typeId -> DefaultActivityEntryTypes.firstOrNull { it.id == typeId } }

val DefaultActivityEntryTypes: List<ActivityEntryType> = listOf(
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_RUNNING,
        labelRes = R.string.exercise_type_running,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_BIKING,
        labelRes = R.string.exercise_type_biking,
        supportsElevation = true,
    ),
    // A trainer ride has a distance and no route; FIT names it (`sub_sport` 5/6).
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_BIKING_STATIONARY,
        id = "stationary_bike",
        labelRes = R.string.exercise_type_stationary_bike,
        supportsGpsRoute = false,
        supportsDistance = true,
        recordingSensor = ActivityRecordingSensor.BLE,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_WALKING,
        labelRes = R.string.exercise_type_walking,
        supportsElevation = true,
        repetitionUnit = ActivityRepetitionUnit.STEPS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_HIKING,
        labelRes = R.string.exercise_type_hiking,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_WHEELCHAIR,
        labelRes = R.string.exercise_type_wheelchair,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_ROWING,
        labelRes = R.string.exercise_type_rowing,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_PADDLING,
        labelRes = R.string.exercise_type_paddling,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SKIING,
        labelRes = R.string.exercise_type_skiing,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SNOWBOARDING,
        labelRes = R.string.exercise_type_snowboarding,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SNOWSHOEING,
        labelRes = R.string.exercise_type_snowshoeing,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SKATING,
        labelRes = R.string.exercise_type_skating,
        supportsElevation = true,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SAILING,
        labelRes = R.string.exercise_type_sailing,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SURFING,
        labelRes = R.string.exercise_type_surfing,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_OPEN_WATER,
        labelRes = R.string.exercise_type_swimming_open_water,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_GOLF,
        labelRes = R.string.exercise_type_golf,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_STRENGTH_TRAINING,
        labelRes = R.string.exercise_type_strength_training,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.BLE,
    ),
    // Generic set sessions for plans that mix exercises. Manual-entry and plan only.
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_STRENGTH_TRAINING,
        id = "strength_sets",
        labelRes = R.string.exercise_type_strength_sets,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.NONE,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_OTHER_WORKOUT,
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS,
        id = "calisthenics",
        labelRes = R.string.exercise_type_calisthenics,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.NONE,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_OTHER_WORKOUT,
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_RUNNING_TREADMILL,
        id = "treadmill",
        labelRes = R.string.exercise_type_treadmill,
        supportsGpsRoute = false,
        supportsDistance = true,
        recordingSensor = ActivityRecordingSensor.STEP_DETECTOR,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_RUNNING_TREADMILL,
        repetitionUnit = ActivityRepetitionUnit.STEPS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS,
        id = "push_ups",
        labelRes = R.string.exercise_type_push_ups,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.PROXIMITY,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_OTHER_WORKOUT,
        defaultTitle = "Push-ups",
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    // Squats have their own segment. Same recognizer as push-ups.
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS,
        id = "squats",
        labelRes = R.string.hc_segment_squat,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.PROXIMITY,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_SQUAT,
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS,
        id = "pull_ups",
        labelRes = R.string.exercise_type_pull_ups,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.ACCELEROMETER,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_PULL_UP,
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS,
        id = "rope_skipping",
        labelRes = R.string.exercise_type_rope_skipping,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.ACCELEROMETER,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_JUMP_ROPE,
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_GYMNASTICS,
        id = "trampoline_jumping",
        labelRes = R.string.exercise_type_trampoline_jumping,
        supportsGpsRoute = false,
        supportsDistance = false,
        recordingSensor = ActivityRecordingSensor.ACCELEROMETER,
        segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_OTHER_WORKOUT,
        defaultTitle = "Trampoline jumping",
        repetitionUnit = ActivityRepetitionUnit.REPETITIONS,
    ),
    ActivityEntryType(
        exerciseType = ExerciseSessionRecord.EXERCISE_TYPE_OTHER_WORKOUT,
        labelRes = R.string.exercise_type_other_workout,
        supportsGpsRoute = false,
        supportsDistance = false,
    ),
)

/**
 * The entry type a plan opens as: a generic set session for mixed plans, or
 * the exercise's own type when one claims the segment.
 */
fun PlannedExerciseData.toActivityEntryType(): ActivityEntryType? {
    val activeSteps = activeSteps()
    if (activeSteps.map { it.exerciseType }.distinct().size > 1) {
        return genericSetActivityType(exerciseType)
    }
    val activeSegmentType = activeSteps
        .firstOrNull { it.completion is PlannedExerciseCompletion.Repetitions }
        ?.exerciseType
    val candidates = DefaultActivityEntryTypes.sortedBy { it.recordingSensor == ActivityRecordingSensor.NONE }
    return candidates.firstOrNull { type ->
        type.exerciseType == exerciseType && type.segmentType != null && type.segmentType == activeSegmentType
    } ?: candidates.firstOrNull { type ->
        type.exerciseType == exerciseType && type.supportsSetRepetitions
    } ?: candidates.firstOrNull { type ->
        type.exerciseType == exerciseType
    }
}

/** The manual-only set type for a session type; strength sets when the type has no generic entry. */
fun genericSetActivityType(exerciseType: Int): ActivityEntryType =
    DefaultActivityEntryTypes.firstOrNull { type ->
        type.exerciseType == exerciseType && type.supportsMixedExercises
    } ?: DefaultActivityEntryTypes.first { it.id == "strength_sets" }

File: lib/domain/model/activity_entry_types.dart
/// Port of the Kotlin `ActivityEntryTypes.kt` (activity type catalogue) plus the
/// Health Connect `ExerciseSessionRecord` / `ExerciseSegment` /
/// `PlannedExerciseStep` constants the manual-entry + recording feature depends
/// on.
///
/// `exerciseType` is the verbatim Health Connect
/// `ExerciseSessionRecord.EXERCISE_TYPE_*` integer constant — the same value the
/// native bridge round-trips through the record JSON (see
/// `health_record_json.dart` and `exercise_labels.dart`), matching the Kotlin
/// source. Exercise-segment types and planned-step phases likewise use verbatim
/// Health Connect integer constants.
class ExerciseSessionType {
  const ExerciseSessionType._();

  static const int running = 56;
  static const int biking = 8;
  static const int bikingStationary = 9;
  static const int walking = 79;
  static const int hiking = 37;
  static const int wheelchair = 82;
  static const int rowing = 53;
  static const int paddling = 46;
  static const int skiing = 61;
  static const int snowboarding = 62;
  static const int snowshoeing = 63;
  static const int skating = 60;
  static const int sailing = 58;
  static const int surfing = 72;
  static const int swimmingOpenWater = 73;
  static const int golf = 32;
  static const int strengthTraining = 70;
  static const int runningTreadmill = 57;
  static const int calisthenics = 13;
  static const int gymnastics = 34;
  static const int otherWorkout = 0;
}

/// Verbatim Health Connect `ExerciseSegment.EXERCISE_SEGMENT_TYPE_*` constants.
class ExerciseSegmentType {
  const ExerciseSegmentType._();

  static const int unknown = 0;
  static const int jumpRope = 28;
  static const int otherWorkout = 38;
  static const int pause = 39;
  static const int pullUp = 42;
  static const int rest = 44;
  static const int runningTreadmill = 47;
}

/// Verbatim Health Connect `PlannedExerciseStep.EXERCISE_PHASE_*` constants.
class PlannedExerciseStepPhase {
  const PlannedExerciseStepPhase._();

  static const int unknown = 0;
  static const int warmup = 1;
  static const int rest = 2;
  static const int active = 3;
}

/// Which live-recording sensor an activity type records from.
enum ActivityRecordingSensor {
  none,
  gps,
  ble,
  proximity,
  accelerometer,
  stepDetector,
}

/// The unit a repetition-like activity counts in.
enum ActivityRepetitionUnit {
  repetitions,
  steps,
}

/// Port of the Kotlin `ActivityEntryType` data class.
class ActivityEntryType {
  ActivityEntryType({
    required this.exerciseType,
    required this.label,
    String? id,
    this.supportsGpsRoute = true,
    this.supportsDistance = true,
    this.supportsElevation = false,
    ActivityRecordingSensor? recordingSensor,
    this.segmentType,
    this.defaultTitle,
    this.repetitionUnit,
  })  : id = id ?? exerciseType.toString(),
        recordingSensor = recordingSensor ??
            (supportsGpsRoute
                ? ActivityRecordingSensor.gps
                : ActivityRecordingSensor.none);

  final int exerciseType;
  final String id;
  final String label;
  final bool supportsGpsRoute;
  final bool supportsDistance;
  final bool supportsElevation;
  final ActivityRecordingSensor recordingSensor;
  final int? segmentType;
  final String? defaultTitle;
  final ActivityRepetitionUnit? repetitionUnit;

  bool get supportsLiveRecording =>
      recordingSensor != ActivityRecordingSensor.none;

  bool get supportsSetRepetitions =>
      repetitionUnit == ActivityRepetitionUnit.repetitions;

  bool get isRepetitionLike => repetitionUnit != null;

  bool get supportsStepCounting =>
      repetitionUnit == ActivityRepetitionUnit.steps;

  @override
  bool operator ==(Object other) =>
      other is ActivityEntryType && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

/// Port of the Kotlin `activityEntryTypeById`.
ActivityEntryType? activityEntryTypeById(String? id) {
  if (id == null) return null;
  for (final type in defaultActivityEntryTypes) {
    if (type.id == id) return type;
  }
  return null;
}

/// Port of the Kotlin `DefaultActivityEntryTypes`.
final List<ActivityEntryType> defaultActivityEntryTypes = <ActivityEntryType>[
  ActivityEntryType(
    exerciseType: ExerciseSessionType.running,
    label: 'Running',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.biking,
    label: 'Cycling',
    supportsElevation: true,
  ),
  // A trainer ride is not a ride that happens to be indoors: it has a distance
  // and no route, and importing it as outdoor cycling would hang a 27 km ride on
  // a map it never touched. FIT names it (`sub_sport` 5/6) and the app had
  // nowhere to put it.
  ActivityEntryType(
    exerciseType: ExerciseSessionType.bikingStationary,
    id: 'stationary_bike',
    label: 'Stationary bike',
    supportsGpsRoute: false,
    supportsDistance: true,
    recordingSensor: ActivityRecordingSensor.ble,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.walking,
    label: 'Walking',
    supportsElevation: true,
    repetitionUnit: ActivityRepetitionUnit.steps,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.hiking,
    label: 'Hiking',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.wheelchair,
    label: 'Wheelchair',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.rowing,
    label: 'Rowing',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.paddling,
    label: 'Paddling',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.skiing,
    label: 'Skiing',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.snowboarding,
    label: 'Snowboarding',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.snowshoeing,
    label: 'Snowshoeing',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.skating,
    label: 'Skating',
    supportsElevation: true,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.sailing,
    label: 'Sailing',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.surfing,
    label: 'Surfing',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.swimmingOpenWater,
    label: 'Open water swimming',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.golf,
    label: 'Golf',
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.strengthTraining,
    label: 'Strength training',
    supportsGpsRoute: false,
    supportsDistance: false,
    recordingSensor: ActivityRecordingSensor.ble,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.runningTreadmill,
    id: 'treadmill',
    label: 'Treadmill',
    supportsGpsRoute: false,
    supportsDistance: true,
    recordingSensor: ActivityRecordingSensor.stepDetector,
    segmentType: ExerciseSegmentType.runningTreadmill,
    repetitionUnit: ActivityRepetitionUnit.steps,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.calisthenics,
    id: 'push_ups',
    label: 'Push-ups',
    supportsGpsRoute: false,
    supportsDistance: false,
    recordingSensor: ActivityRecordingSensor.proximity,
    segmentType: ExerciseSegmentType.otherWorkout,
    defaultTitle: 'Push-ups',
    repetitionUnit: ActivityRepetitionUnit.repetitions,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.calisthenics,
    id: 'pull_ups',
    label: 'Pull-ups',
    supportsGpsRoute: false,
    supportsDistance: false,
    recordingSensor: ActivityRecordingSensor.accelerometer,
    segmentType: ExerciseSegmentType.pullUp,
    repetitionUnit: ActivityRepetitionUnit.repetitions,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.calisthenics,
    id: 'rope_skipping',
    label: 'Rope skipping',
    supportsGpsRoute: false,
    supportsDistance: false,
    recordingSensor: ActivityRecordingSensor.accelerometer,
    segmentType: ExerciseSegmentType.jumpRope,
    repetitionUnit: ActivityRepetitionUnit.repetitions,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.gymnastics,
    id: 'trampoline_jumping',
    label: 'Trampoline jumping',
    supportsGpsRoute: false,
    supportsDistance: false,
    recordingSensor: ActivityRecordingSensor.accelerometer,
    segmentType: ExerciseSegmentType.otherWorkout,
    defaultTitle: 'Trampoline jumping',
    repetitionUnit: ActivityRepetitionUnit.repetitions,
  ),
  ActivityEntryType(
    exerciseType: ExerciseSessionType.otherWorkout,
    label: 'Other workout',
    supportsGpsRoute: false,
    supportsDistance: false,
  ),
];
